New engine!

chrisnojima opened the pull request.

chrisnojima commented on desktop/app/index.js:
```diff
@@ -1,7 +1,6 @@
 import {BrowserWindow, app, dialog} from 'electron'
 import splash from './splash'
 import installer from './installer'
-import ListenLogUi from '../shared/native/listen-log-ui'
```
this is now handled in the engine itself

chrisnojima commented on protocol/Makefile:
```diff
@@ -33,18 +33,13 @@ js/flow-types.js: build-stamp | config
 	node --harmony ./bin/flow.js
 	cp ./js/flow-types* ../shared/constants/types
 
-js/keybase_v1.js: build-stamp | config
```
the enums have moved to flow-types. Also reduced the hierarchy a level so instead of

```
common.StatusCodes.scgeneric
```

you do

```
CommonStatusCodes.scgeneric
```

chrisnojima commented on protocol/bin/flow.js:
```diff
@@ -46,31 +49,74 @@ function load (file, project) {
+  return s.toLowerCase().replace(/(_\w)/g, s => capitalize(s[1]))
+}
+
+function analyzeEnums (json, project) {
```
handle enums now

chrisnojima commented on protocol/bin/flow.js:
```diff
@@ -92,6 +138,9 @@ function capitalize (s) {
 
 function analyzeMessages (json, project) {
+  // ui means incoming
+  const isUIProtocol = ['notifyCtl'].indexOf(json.protocol) === -1 && !!json.protocol.match(/^notify.*|.*ui$/i)
```
we keep track of whats a ui call or not. this means we can filter out the outgoing calls (the ...rpc()) and the incomingCallMap into the 2 types.

chromakode replied:
Perhaps this regex shouldn't be case-insensitive?

chrisnojima replied:
it has to be to be correct

chrisnojima replied:
i'll add a comment

chrisnojima commented on protocol/bin/flow.js:
```diff
@@ -254,10 +312,31 @@ export type RPCError = {
-  Object.keys(project.incomingMaps).map(im => `  '${im}'?: ${project.incomingMaps[im]}`).join(',\n') + '\n}>\n\n'
-  s.write(typePrelude + typeDefs.join('') + incomingMap)
+  Object.keys(project.incomingMaps).map(im => `  '${im}'?: ${project.incomingMaps[im]}`).join(',\n') + '\n}>\n'
+  const toWrite = [typePrelude, typeDefs.join('\n\n'), incomingMap].join('\n')
```
do our newlines here to control the spacing easier

chrisnojima commented on protocol/bin/flow.js:
```diff
@@ -254,10 +312,31 @@ export type RPCError = {
   desc: string
 }
-`
+export type WaitingHandlerType = (waiting: boolean, method: string, sessionID: number) => void
```
use more helper types here to reduce the amount of code we produce

chrisnojima commented on shared/.flowconfig:
```diff
@@ -33,6 +33,7 @@
 munge_underscores=true
 esproposal.class_static_fields=enable
 esproposal.class_instance_fields=enable
+unsafe.enable_getters_and_setters=true
```
engine uses some getters/setters

chrisnojima commented on shared/actions/gregor.js:
```diff
@@ -48,11 +48,9 @@ function registerGregorListeners () {
-        dispatch(pushState(state, reason))
-        response && response.result && response.result()
-      },
+    engine.setIncomingHandler('keybase.1.gregorUI.pushState', ({state, reason}, response) => {
```
This method takes in a single handler and not an object of them

chrisnojima commented on shared/actions/login/index.js:
```diff
@@ -99,7 +100,7 @@ export function login (): AsyncAction {
             deviceType,
             usernameOrEmail: usernameOrEmail,
-            clientType: Common.ClientType.gui,
+            clientType: CommonClientType.gui,
```
new enums format

chrisnojima commented on shared/actions/login/index.js:
```diff
@@ -110,7 +111,7 @@ export function login (): AsyncAction {
               })
 
-              if (!(error.raw && error.raw.code === InputCancelError.code)) {
+              if (!(error.code === InputCancelError.code)) {
```
removed the .raw stuff

chrisnojima commented on shared/actions/tracker.js:
```diff
@@ -411,11 +416,11 @@ function updatePGPKey (username: string, pgpFingerprint: Buffer): Action {
-  const sessionIDToUsername: { [key: number]: string } = {}
-  const identifyUi = {
-    start: ({username, sessionID, reason}) => {
+function serverCallMap (dispatch: Dispatch, getState: Function, skipPopups: boolean = false, onFinish: ?() => void): incomingCallMapType {
```
removed the call map middleware. this makes this look like everything else and also flowtypes better

chrisnojima commented on shared/actions/unlock-folders.js:
```diff
@@ -46,21 +41,13 @@ export function finish (): Finish {
-      method: 'rekey.showPendingRekeyStatus',
-      callback: null,
-    }
-    engine.rpc(params)
```
dunno why this old style call exists

chrisnojima commented on shared/actions/unlock-folders.js:
```diff
@@ -46,21 +41,13 @@ export function finish (): Finish {
-    rekeyRekeyStatusFinishRpc({
-      param: {sessionID: rekeyServer && rekeyServer.sessionID},
-    })
+    rekeyRekeyStatusFinishRpc({})
```
with the rekey refactor on core we don't need to bookkeep the sessionID anymore

chrisnojima commented on shared/actions/unlock-folders.js:
```diff
@@ -80,19 +67,17 @@ export function registerRekeyListener (): (dispatch: Dispatch) => void {
-      null,
-      () => ({
-        'keybase.1.rekeyUI.delegateRekeyUI': (params, response) => { },
+    engine.setIncomingHandler('keybase.1.rekeyUI.delegateRekeyUI', (param: any, response: ?Object) => {
```
instead of the whole server layer on top of engine we simply listen for this message and create a new session

chrisnojima commented on shared/engine/errors.js:
```diff
@@ -1,41 +0,0 @@
-// @flow
```
not wrapping this stuff anymore. instead actions/reducers should deal with error code transformations

chrisnojima commented on shared/engine/index.js:
```diff
@@ -1,430 +1,284 @@
 // @flow
```
This has changed a ton and should be looked at as a non-diff

chrisnojima replied:
tldr; Before the engine had maps to keep track of session ids, responses, outstanding requests, etc. Now it just knows about sessions and generalized handlers mostly

chrisnojima commented on shared/engine/request.js:
```diff
@@ -0,0 +1,99 @@
+// @flow
```
Sessions have requests happen

chrisnojima commented on shared/engine/transport-shared.js:
```diff
@@ -1,14 +1,85 @@
 // @flow
-import EngineError from './errors'
+// Classes used to handle RPCs. Ability to inject delays into calls to/from server
```
delay, logging, only once behavior is now all at this level

chrisnojima commented:
@keybase/react-hackers bump

chrisnojima commented:
@keybase/react-hackers bump please

MarcoPolo commented on shared/engine/index.js:
```diff
@@ -1,435 +1,284 @@
+      log(logParam)
+      response && response.result && response.result()
     })
+    this.setIncomingHandler('keybase.1.NotifyUsers.userChanged', () => {})
```
We should be doing something with this in the future, right?

chrisnojima replied:
unclear

chromakode replied:
We should add a comment to this line explaining why it's no-opped

chrisnojima replied:
done, moved to its own _setupIgnoredHanlders  handler to make things more clear

MarcoPolo commented on shared/engine/index.js:
```diff
@@ -1,435 +1,284 @@
+      this._handleCancel(seqid)
+    } else {
+      const session = this._sessionsMap[String(sessionID)]
+      if (session && session.incomingCall(method, param, response)) { // Part of a session?
```
Just a note for myself and other reviewers:

This lets the session say "I can't handle this" even if there is a sesssion that matches that session id.

Useful for stuff like pinentry where we don't want each session handling it. It's okay to just fallback

chromakode commented on protocol/bin/flow.js:
```diff
@@ -26,18 +26,21 @@ var projects = [
   },
 ]
 
+const reduceArray = arr => arr.reduce((acc, cur) => acc.concat(cur), [])
```
https://lodash.com/docs#flatten ?

MarcoPolo commented:
not for this PR, but in the future we might want rpc functions able to return something that can be used to cancel them later.

MarcoPolo commented on protocol/js/flow-types-chat.js:
```diff
@@ -16,6 +16,60 @@ export type RPCError = {
+export type WaitingHandlerType = (waiting: boolean, method: string, sessionID: number) => void
+
+// $FlowIssue we're calling an internal method on engine that's there just for us
+const engineRpcOutgoing = (...args) => engine._rpcOutgoing(...args)
```
This is weird that this imports engine for this. Maybe it would be better to have engine be the first argument when you make rpc calls?

It's double weird when you think that importing engine has the side effect of creating a new engine instance. We should probably export a singleton constructor instead.

In the singleton constructor, it doesn't make sense to build the engine here, because that would mean that anytime you want to use types you are building an engine.

I know @gabriel has a couple thoughts about this too.

chrisnojima replied:
so this code already existed. (not that we can't change it). this is so people who are making rpc calls don't need to understand that engine exists else they'll all have to import it just to pass it into this helper (which seems lame).
new engines are not being made, its a singleton thats exported by the module. if we want to export a closure that also exports a singleton thats fine but that isn't a fix for what you're talking about

gabriel replied:
Yeah, I listed some thoughts on that here:
https://github.com/keybase/client/pull/4060

I think constructing (non-simple things) via import and using import to enforce singleton-ness is worth avoiding.

Also especially in this case because engine construction I think triggers an rpc connection.

Not a huge change to make this happen, we can do it after in a separate PR.

chrisnojima replied:
i buy controlling when its constructed. we can force you to do engine() like you suggest. i'm just saying we're not building multiple engines

chrisnojima replied:
also node framed rpc doens't support sending cancels from our side currently so let's not try and make that work short term

MarcoPolo replied:
I didn't mean we were building multiple engines. I meant that importing a type has a side effect of building an engine (assuming it didn't exist already).

chromakode commented on protocol/bin/flow.js:
```diff
@@ -46,31 +49,74 @@ function load (file, project) {
+    }
+  }).reduce((acc, t) => {
+    return acc.concat([
+      `export const ${t.name} = {
```
:sheep: `dedent` is pretty nice for this https://www.npmjs.com/package/dedent-js

chrisnojima replied:
protocol has its own deps and isn't something we touch that often so i'm trying to not have too many dependencies

MarcoPolo commented on shared/engine/session.js:
```diff
@@ -0,0 +1,157 @@
+            this._seqIDResponded[String(seqid)] = true
+          }
+          // Remove from our list
+          requests.splice(idx, 1)
```
not a fan of mutation...

what about 

```
if (isOutgoing) {
 this._outgoingReqs.filter(r => r.method !== method)
} else {
//same
}

if (seqid) {
  this._seqIDResponded[String(seqid)] = true
}
```

chrisnojima replied:
i think we want this stuff cleaned up, its our own internal bookkeeping so why have it hang around?

MarcoPolo replied:
actually I see you probably want seqIDResponded only true if you had this in your known requests.   So tweak that demo a bit

MarcoPolo replied:
This is a 🐑

MarcoPolo commented on shared/engine/session.js:
```diff
@@ -0,0 +1,157 @@
+  // Start the session normally. Tells engine we're done at the end
+  start (method: MethodKey, param: ?Object, callback: ?() => void) {
+    this._startMethod = method
+    this._startCallback = callback
```
🐑 🐑 `startCallback` is a bit confusing of a name, but I can't think of a better one

chromakode commented on protocol/bin/flow.js:
```diff
@@ -217,28 +269,31 @@ function parseRecord (t) {
+    const m = t.match(/(\w*Rpc) \(/)
+    return m && m[1]
+  })
+  .filter(t => t)
```
Filter no longer required here?

chrisnojima replied:
still needed actually

chromakode replied:
Ah I see -- duh!

MarcoPolo commented on shared/engine/transport-shared.js:
```diff
@@ -24,7 +98,38 @@ class TransportShared extends RobustTransport {
+
+      calls.forEach(call => {
+        payload.response[call] = _wrap((...args) => {
+          oldResponse[call](...args)
```
so you don't need the context behind `response.result` anymore?

we can do destructuring in callbacks now?

i.e. `({response: {result}) => result()`

chrisnojima replied:
i think this is a dangerous pattern and should be avoided. the caller changes binding and now you're broken again

MarcoPolo replied:
you mean `this` is a dangerous pattern ;)

chrisnojima replied:
yeah sadly. the payloads are plain-ol-javascript-objects, but response isn't so destructuring it is gunna be dangerous.

MarcoPolo commented:
👍 Do I get a 🍪 ?

MarcoPolo commented:
👍

chromakode commented on shared/actions/pinentry.js:
```diff
@@ -28,8 +27,28 @@ export function registerPinentryListener (): AsyncAction {
 
-    const pinentryListeners = pinentryListenersCreator(dispatch)
-    engine.listenGeneralIncomingRpc(pinentryListeners)
+    engine.setIncomingHandler('keybase.1.secretUi.getPassphrase', (payload, response) => {
```
:sheep: Might be valuable to give this function a name (rather than arrow function) for stack traces

chrisnojima replied:
we don't really do this elsewhere. also with the current stack you can see its a getPassphrase call and thats its a handler so i think its all easy to understand

chromakode commented on shared/engine/index.js:
```diff
@@ -281,4 +279,12 @@ export type ResponseType = {
 
-export default engine
+let engine
+const makeEngine = () => {
```
I think this is an improvement on what was before, but one drawback is that the timing of creating the engine is implicit based on first use. This approach introduces a global but defers the organizational task of deciding where to fit it into the namespace and when to initialize it. This is technical debt.

Do we have some other place where shared globals are setup / stored? Should we create a pattern for this?

chromakode replied:
Alternatively a more explicit `setupEngine` called in app startup would be a welcome addition.

chrisnojima replied:
i added this. we just init it when we're setting up a new context

chromakode commented on shared/engine/index.js:
```diff
@@ -1,435 +1,290 @@
+    param?: ?Object,
+    incomingCallMap?: incomingCallMapType,
+    callback?: ?(...args: Array<any>) => void,
+    waitingHandler?: WaitingHandlerType}) {
```
It's pretty hard to tell where the arguments end and the functions begin with this style. Maybe drop the closing `) {` to the next line and dedent?

chromakode commented:
Epic! :+1: % style comments

chrisnojima commented:
@MarcoPolo @chromakode https://github.com/keybase/client/pull/3953/commits/c71e3070f3da99db9f3c716f7e34c786ab966a03 addresses your pr comments and then i rebased after. can i get one more thumb then i'll merge

chromakode commented on shared/engine/index.js:
```diff
@@ -287,10 +289,30 @@ export type ResponseType = {
+
+  // This is just a sanity check so we don't break old code. Should never happen in practice
+  if (!engine) {
+    makeEngine()
```
Should we make this throw in dev?

chromakode replied:
Ah I see, handled by `makeEngine` above. NVM!

chromakode commented:
Updates look good! :+1:

Re: `makeEngine` calls, I didn't notice any for the mobile code. Is that TBD?

chrisnojima commented:
no lemme fix that tnx

chrisnojima commented:
sadly i have to update react/react-native on the rn side so we can get pure components for the profile updates

chrisnojima commented:
@chromakode @MarcoPolo updated w/ mobile fix

chromakode commented:
:+1:

chromakode commented:
![](http://vignette3.wikia.nocookie.net/adventuretimewithfinnandjake/images/a/a9/Adventure-time-dance-5046_preview.gif/revision/latest?cb=20110617014146)
